### PR TITLE
Add generic IDictionary tests to ConcurrentDictionary

### PR DESCRIPTION
**Pull request by hughbe**
- Ensure ConcurrentDictionary meets the spec for IDictionary objects

I had to add quite a lot of customization points, but I don't think these are due to bugs

/cc @ianhays

### CODE REVIEW

File: src/Common/tests/System/Collections/IDictionary.Generic.Tests.cs [hunk 1]
```diff
@@ -193,6 +193,24 @@ protected override IEnumerable<ModifyEnumerable> ModifyEnumerables
...
+        /// Used in IDictionary_Generic_Values_Enumeration_ParentDictionaryModifiedInvalidates and 
+        /// IDictionary_Generic_Keys_Enumeration_ParentDictionaryModifiedInvalidates.
+        /// Some collections (e.g. ConcurrentDictionary) do not throw an InvalidOperationException
+        /// when enumerating the Keys or Values property and the parent is modified.
+        /// </summary>
+        protected virtual bool IDictionary_Generic_Keys_Values_Enumeration_ThrowsInvalidOperation_WhenParentModified => true;
+
+        /// <summary>
+        /// Used in IDictionary_Generic_Values_ModifyingTheDictionaryUpdatesTheCollection and 
+        /// IDictionary_Generic_KKeyss_ModifyingTheDictionaryUpdatesTheCollection/
```
Comment (ianhays) [inline src/Common/tests/System/Collections/IDictionary.Generic.Tests.cs]: nit: KKeyss

File: src/Common/tests/System/Collections/IDictionary.Generic.Tests.cs [hunk 2]
```diff
@@ -323,23 +341,34 @@ public void IDictionary_Generic_Keys_ModifyingTheDictionaryUpdatesTheCollection(
...
         public void IDictionary_Generic_Keys_Enumeration_ParentDictionaryModifiedInvalidates(int count)
         {
-            IDictionary<TKey, TValue> dictionary = GenericIDictionaryFactory(count);
-            ICollection<TKey> keys = dictionary.Keys;
-            IEnumerator<TKey> keysEnum = keys.GetEnumerator();
-            dictionary.Add(GetNewKey(dictionary), CreateTValue(3432));
-            Assert.Throws<InvalidOperationException>(() => keysEnum.MoveNext());
-            Assert.Throws<InvalidOperationException>(() => keysEnum.Reset());
-            var cur = keysEnum.Current;
+            if (IDictionary_Generic_Keys_Values_Enumeration_ThrowsInvalidOperation_WhenParentModified)
```
Comment (ianhays) [inline src/Common/tests/System/Collections/IDictionary.Generic.Tests.cs]: I sure wish we could use ConditionalFacts/ConditionalTheories on non-static methods.

File: src/Common/tests/System/Collections/IEnumerable.Generic.Tests.cs [hunk 3]
```diff
@@ -64,6 +64,18 @@ public abstract class IEnumerable_Generic_Tests<T> : TestBase<T>
...
+        /// Used in IEnumerable_Generic_Enumerator_Current.
+        /// Some enumerators do not throw accessing Current if enumeration has not yet started (e.g. ConcurrentDictionary)
+        /// </summary>
+        protected virtual bool IEnumerable_Generic_Enumerator_Current_EnumerationNotStarted_ThrowsInvalidOperationException => true;
+
+        /// <summary>
+        /// Some collections (e.g. ConcurrentDictionary) use yield return iterators to enumerate,
+        /// so calling reset anytime throws a NotSupportedException.
+        /// </summary>
+        protected virtual bool IEnumerable_Generic_Enumerator_Reset_Implemented => false;
```
Comment (ianhays) [inline src/Common/tests/System/Collections/IEnumerable.Generic.Tests.cs]: How is this different from ```ResetImplemented```?

**Comment by ianhays** [comment]
Thanks for updating these, Hugh! Just have a few nits.

File: src/Common/tests/System/Collections/IEnumerable.Generic.Tests.cs (hunk 3 above)
Comment (hughbe) [inline src/Common/tests/System/Collections/IEnumerable.Generic.Tests.cs]: It's dead code that I found its way in when I was diagnosing why ResetImplemented also assumes that Reset is implemented or not in the keys or values collections of an IDictionary. Will remove

**Comment by hughbe** [comment]
Thanks for the review, I've addressed your comments

**Comment by ianhays** [comment]
LGTM!

**Comment by stephentoub** [comment]
LGTM2

**Pull request merged**